libs/langchain-google-common/src/experimental/utils/media_core.ts

```diff
@@ -0,0 +1,504 @@
+    const suffix = this._blobPathSuffix(blob);
+    const newPath = `${prefix}${suffix}`;
+    return this._newBlob(blob, newPath);
+  }
+
+  protected async _validBlobPrefixUuid(
+    blob: MediaBlob,
+    opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob> {
+    const prefix = opts?.pathPrefix ?? "";
+    const suffix = uuidv4(); // TODO - option to specify version?
+    const newPath = `${prefix}${suffix}`;
+    return this._newBlob(blob, newPath);
+  }
+
+  protected async _validBlobRemovePath(
+    blob: MediaBlob,
+    _opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob> {
+    return this._newBlob(blob, undefined);
+  }
+
+  /**
+   * Based on the blob and options, return a blob that has a valid path
+   * that can be saved.
+   * @param blob
+   * @param opts
+   */
+  protected async _validStoreBlob(
+    blob: MediaBlob,
+    opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob | undefined> {
+    if (await this._hasValidPath(blob, opts)) {
+      return blob;
+    }
+    switch (opts?.actionIfInvalid) {
+      case "ignore":
+        return blob;
+      case "prefixPath":
+        return this._validBlobPrefixPath(blob, opts);
+      case "prefixUuid":
+        return this._validBlobPrefixUuid(blob, opts);
+      case "removePath":
+        return this._validBlobRemovePath(blob, opts);
+      default:
+        return undefined;
+    }
+  }
+
+  async store(
+    blob: MediaBlob,
+    opts: BlobStoreStoreOptions = {}
+  ): Promise<MediaBlob | undefined> {
+    const allOpts: BlobStoreStoreOptions = {
+      ...this.defaultStoreOptions,
+      ...opts,
+    };
+    const validBlob = await this._validStoreBlob(blob, allOpts);
+    if (typeof validBlob !== "undefined") {
+      const validKey = await validBlob.asUri();
+      await this.mset([[validKey, validBlob]]);
+      const savedKey = await validBlob.asUri();
+      return await this.fetch(savedKey);
+    }
+    return undefined;
+  }
+
+  protected async _missingFetchBlobEmpty(
+    path: string,
+    _opts?: BlobStoreFetchOptions
+  ): Promise<MediaBlob> {
+    return new MediaBlob({ path });
+  }
+
+  protected async _missingFetchBlob(
+    path: string,
+    opts?: BlobStoreFetchOptions
+  ): Promise<MediaBlob | undefined> {
+    switch (opts?.actionIfBlobMissing) {
+      case "emptyBlob":
+        return this._missingFetchBlobEmpty(path, opts);
+      default:
+        return undefined;
+    }
+  }
+
+  async fetch(
+    key: string | MediaBlob,
+    opts: BlobStoreFetchOptions = {}
+  ): Promise<MediaBlob | undefined> {
+    const allOpts: BlobStoreFetchOptions = {
+      ...this.defaultFetchOptions,
+      ...opts,
+    };
+    const realKey = await this._realKey(key);
+    const ret = await this.mget([realKey]);
+    return ret?.[0] ?? (await this._missingFetchBlob(realKey, allOpts));
+  }
+}
+
+export interface BackedBlobStoreOptions extends BlobStoreOptions {
+  backingStore: BaseStore<string, MediaBlob>;
+}
+
+export class BackedBlobStore extends BlobStore {
```

Comment:
Yes.

I use it in tests for `MediaManager` for an `aliasStore` that is backed by a memory store.
I can also see it used in general for either the `aliasStore` or a retriever that will be using some other defined store to store things. I don't want people to need to re-invent the storage methodology to support the `BlobStore` semantics, but I do want to be able to support them.

Reply:
I don't really understand this one - why not just interact with `this.backingStore` directly?

If just used for tests, put in a `utils/testing` file

Reply:
Because the `BlobStore` versions have different semantics for storing an object (they return values from a set while the `BaseStore` versions don't). So rather than reimplementing all of the `BaseStore` implementations (now and in the future), it guarantees those semantics for other `BaseStore` types.  A memory store being the biggest example, but anything that may save to a database or whatever also makes sense.